src/dft/src/ScanReplace.cpp

```diff
@@ -0,0 +1,426 @@
+    if (non_scan_cell_port->function() == nullptr
+        && scan_cell_port->function() == nullptr) {
+      // input ports do not have a function
+      port_equiv
+          = port_equiv
+            && strcmp(non_scan_cell_port->name(), scan_cell_port->name()) == 0;
+    } else {
+      port_equiv = port_equiv
+                   && sta::FuncExpr::equiv(non_scan_cell_port->function(),
+                                           scan_cell_port->function());
+    }
+
+    if (port_equiv) {
+      return scan_cell_port;
+    }
+  }
+  return nullptr;
+}
+
+// Checks the power ports between non scan cell and scan cell and checks if they
+// are equivalent. Returns the equivalent port found, otherwise nullptr if there
+// is none
+sta::LibertyPgPort* FindEquivalentPortInScanCell(
+    const sta::LibertyPgPort* non_scan_cell_port,
+    const sta::LibertyCell* scan_cell)
+{
+  sta::LibertyCellPgPortIterator scan_cell_ports_iter(scan_cell);
+  while (scan_cell_ports_iter.hasNext()) {
+    sta::LibertyPgPort* scan_cell_port = scan_cell_ports_iter.next();
+    const bool port_equiv
+        = sta::LibertyPgPort::equiv(non_scan_cell_port, scan_cell_port);
+    if (port_equiv) {
+      return scan_cell_port;
+    }
+  }
+  return nullptr;
+}
+
+// We look for each of the non_scan_cell's ports for a functional equivalent one
+// in the scan_cell. If there is a port that we are unable to find an scan
+// equivalent, then we can say that the given scan cell is not a replacement for
+// the given non-scan one
+bool IsScanEquivalent(
+    const sta::LibertyCell* non_scan_cell,
+    const sta::LibertyCell* scan_cell,
+    std::unordered_map<std::string, std::string>& port_mapping)
+{
+  sta::LibertyCellPortIterator non_scan_cell_ports_iter(non_scan_cell);
+  while (non_scan_cell_ports_iter.hasNext()) {
+    sta::LibertyPort* non_scan_cell_port = non_scan_cell_ports_iter.next();
+    sta::LibertyPort* scan_equiv_port
+        = FindEquivalentPortInScanCell(non_scan_cell_port, scan_cell);
+    if (!scan_equiv_port) {
+      return false;
+    } else {
+      port_mapping.insert(
+          {non_scan_cell_port->name(), scan_equiv_port->name()});
+    }
+  }
+
+  sta::LibertyCellPgPortIterator non_scan_cell_pg_ports_iter(non_scan_cell);
+  while (non_scan_cell_pg_ports_iter.hasNext()) {
+    sta::LibertyPgPort* non_scan_cell_pg_port
+        = non_scan_cell_pg_ports_iter.next();
+    sta::LibertyPgPort* scan_equiv_port
+        = FindEquivalentPortInScanCell(non_scan_cell_pg_port, scan_cell);
+    if (!scan_equiv_port) {
+      return false;
+    } else {
+      port_mapping.insert(
+          {non_scan_cell_pg_port->name(), scan_equiv_port->name()});
+    }
+  }
+
+  return true;
+}
+
+// We calculate the difference in performance of the ports given
+double DifferencePerformancePorts(const sta::LibertyPort* port1,
+                                  const sta::LibertyPort* port2)
+{
+  double diff = 0;
```

Comment:
adding incompatible units

Reply:
Removed. Solved in DifferencePerformanceCells